iterator_test.go

```diff
@@ -2737,3 +2737,176 @@ func BenchmarkSeekPrefixTombstones(b *testing.B) {
+		n := testkeys.WriteKey(key, queueIDKeyspace, int64(q))
+		key[n] = '/'
+		prefixLen := n + 1
+		n = testkeys.WriteKey(key[prefixLen:], itemKeyspace, int64(i))
+		return key[:prefixLen+n]
+	}
+
+	type Queue struct {
+		start int
+		end   int // exclusive
+	}
+	var queues = make([]*Queue, queueCount)
+	for i := 0; i < queueCount; i++ {
+		queues[i] = &Queue{}
+	}
+
+	o := (&Options{
+		DisableWAL:         true,
+		FS:                 vfs.NewMem(),
+		Comparer:           testkeys.Comparer,
+		FormatMajorVersion: FormatNewest,
+	}).EnsureDefaults()
+
+	d, err := Open("", o)
+	require.NoError(b, err)
+
+	processQueueOnce := func(batch *Batch, batchSize int) *Batch {
+		for {
+			// Randomly pick a queue to process.
+			q := rng.Intn(queueCount)
+			queue := queues[q]
+
+			isDelete := rng.Float32() < deleteRatio
+
+			if isDelete {
+				// Only process the queue if it's not empty. Otherwise, retry
+				// with a different queue.
+				if queue.start < queue.end {
+					require.NoError(b, batch.Delete(getKey(q, queue.start), nil))
+					queue.start++
+					break
+				}
+			} else {
+				// Append to the queue.
+				require.NoError(b, batch.Set(getKey(q, queue.end), val, nil))
+				queue.end++
+				break
+			}
+		}
+
+		if batch.Len() >= batchSize {
+			require.NoError(b, batch.Commit(NoSync))
+			d.AsyncFlush()
+			return d.NewBatch()
+		}
+		return batch
+	}
+
+	// First, process queues initialOps times.
+	batch := d.NewBatch()
+	for i := 0; i < initOps; i++ {
+		// Use a large batch size to speed up initialization.
+		batch = processQueueOnce(batch, 10<<24 /* 167 MiB */)
+	}
+	require.NoError(b, batch.Commit(NoSync))
+
+	b.Run(fmt.Sprintf("initial_ops=%d,deleteRatio=%.2f,valueSize=%d", initOps, deleteRatio, valueSize), func(b *testing.B) {
+		waitForCompactions := func() {
+			d.mu.Lock()
+			// NB: Wait for table stats because some compaction types rely
+			// on table stats to be collected.
+			d.waitTableStats()
+			for d.mu.compact.compactingCount > 0 {
+				d.mu.compact.cond.Wait()
+				d.waitTableStats()
+			}
+			d.mu.Unlock()
+		}
+
+		waitForCompactions()
+
+		// Log the number of tombstones and live keys in each level after
+		// background compactions are complete.
+		b.Log("LSM after compactions:")
+		firstIter, _ := d.NewIter(nil)
+		firstIter.First()
+		lastIter, _ := d.NewIter(nil)
+		lastIter.Last()
+		stats, _ := d.ScanStatistics(context.Background(), firstIter.Key(), lastIter.Key(), ScanStatisticsOptions{})
+		require.NoError(b, firstIter.Close())
+		require.NoError(b, lastIter.Close())
+		metrics := d.Metrics()
+		for i := 0; i < numLevels; i++ {
+			numTombstones := stats.Levels[i].KindsCount[base.InternalKeyKindDelete]
+			numSets := stats.Levels[i].KindsCount[base.InternalKeyKindSet]
+			numTables := metrics.Levels[i].NumFiles
+			if numSets > 0 {
+				b.Logf("L%d: %d tombstones, %d sets, %d sstables\n", i, numTombstones, numSets, numTables)
+			}
+		}
+
+		b.Run("queue_throughput", func(b *testing.B) {
```

Comment:
Since this part is just batch applications, which are basically a memtable (+ WAL write except wals are disabled), I don't see how it will be affected by any compaction heuristics or really any parameters. We can probably move this outside of b.Run, and also remove `seek` and just time the part currently within `seek` in the main benchmark.

The other reason not to time this part is that if we get unlucky with the randomization in `processQueueOnce`, and the deleteRatio is high, we could end up looping around the queues a bunch of times before we find one that has an object to delete.

Reply:
My goal with this sub-benchmark was to surfaces cases where (a) queue throughput slows down as tombstones build up, and (b) throughput slows down due to background compactions that are scheduled as a result of any heuristic that I add. For (a), currently it looks like there is a clear throughput slowdown as `initialOps` increases for small value sizes, but for large values sizes the numbers are pretty random, so I agree that the usefulness is questionable.

For (b), I'm curious about your thoughts on whether this is a good way to test for performance hits due to excessive background compactions. One other idea I had was to time `waitForCompactions`, but this doesn't fit into the benchmarking model of repeating `b.N` times.

I've removed this for now since seek time is definitely the main number we care about, but open to thoughts on alternative ways to test for (b).

Reply:
Update: I added back the `b.Run("seek", ...` subbenchmark so that the DB is only initialized once per configuration. Without it, the whole `runQueueBenchmarkWorkload` helper is called multiple times as the benchmark increases `b.N`. We can still test with multiple initializations to account for randomness using the `-count` flag.

Reply:
LSMs are inherently going to get as many things out of the write path as possible, and compactions are a background / nondeterministic process so benchmarking them directly is more challenging. I don't see any cases where queue throughput (as long as that only includes writes) would be slowed down by any buildup, unless you're hitting write stall thresholds which will be pretty hard.

So I think it still makes sense to not have a `queue_throughput` sub-benchmark, but one just for seeks is understandable. This sounds good, thanks!